fe/fe-core/src/main/java/com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java

```diff
@@ -102,6 +110,51 @@ public static Type normalizeDecimalArgTypes(Type[] argTypes, FunctionName fnName
         return Type.INVALID;
     }
 
+    public static Type getReturnTypeOfTruncate(FunctionCallExpr node, List<Type> argTypes) {
+        Type firstArgType = argTypes.get(0);
+        Expr secondArg = node.getParams().exprs().get(1);
+
+        // TODO(hcf) How to deduce the type of return
+        if (!firstArgType.isDecimalV3()) {
+            return Type.DEFAULT_DECIMAL128;
+        }
+
+        // For simplicity, we use decimal128(38, ?) as return type, so we only need to
+        // figure out the scale
+        final int originalScale = ((ScalarType) firstArgType).getScalarScale();
+        final PrimitiveType returnPrimitiveType = PrimitiveType.DECIMAL128;
+        final int returnPrecision = PrimitiveType.getMaxPrecisionOfDecimal(PrimitiveType.DECIMAL128);
+        final int returnScale;
+
+        if (secondArg instanceof IntLiteral) {
+            final int expectedScale = (int) ((IntLiteral) secondArg).getValue();
+
+            // If scale expand, we use the maximum precision as the result's precision
+            if (expectedScale > originalScale) {
+                // truncate(0.1, 10000) is treated as truncate(0.1, 38), type of result is decimal128(38, 38)
+                if (expectedScale > PrimitiveType.getMaxPrecisionOfDecimal(PrimitiveType.DECIMAL128)) {
+                    returnScale = PrimitiveType.getMaxPrecisionOfDecimal(PrimitiveType.DECIMAL128);
+                } else {
+                    returnScale = expectedScale;
+                }
+            } else if (expectedScale < 0) {
+                // Invalid expectedScale, modify it to lower bounder
+                returnScale = 0;
+            } else {
+                // Scale reduce
+                returnScale = expectedScale;
+            }
+        } else if (Expr.containsSlotRef(secondArg)) {
```

Comment:
should call !isConstantImpl?

Reply:
We decide not support constant expression, because not all the constant can be calcuated in fe